src/core/manager.c

```diff
@@ -1250,6 +1250,31 @@ static unsigned manager_dispatch_cleanup_queue(Manager *m) {
         return n;
 }
 
+static unsigned manager_dispatch_cgroup_empty_queue(Manager *m) {
+        Unit *u;
+        unsigned n = 0;
+
+        assert(m);
+
+        while ((u = m->cgroup_empty_queue)) {
+                assert(u->in_cgroup_empty_queue);
+
+                u->in_cgroup_empty_queue = false;
+                LIST_REMOVE(cgroup_empty_queue, m->cgroup_empty_queue, u);
+
+                /* Update state based on OOM kills before we notify about cgroup empty event */
+                (void) unit_check_oom(u);
+                (void) unit_check_oomd_kill(u);
+
+                unit_add_to_gc_queue(u);
+
+                if (UNIT_VTABLE(u)->notify_cgroup_empty)
+                        UNIT_VTABLE(u)->notify_cgroup_empty(u);
+        }
+
+        return n;
+}
```

Comment:
this looks wrong. now there are two places where the cgroup empty queue is dispatched, here, synchronously from the event loop, and very similar in style from a low-prio defer event source. the low-prefere defer event source exists so that we can let it run *after* the SIGCHLD handler.

If you want to ensure the GC logic runs only after cgroup empty events have been handled for a unit, then modify unit_may_gc() maybe to make sure that units are never GC'ed which are in the cgroup empty event queue.

Reply:
or in other words, i want that on_cgroup_empty_event() remains the only place where we dispatch the cgroup empty event queue – however, to not have the GC logic interfere, let's make it ineffective if cgroup empty event is pending in the queue for the unit.